### ISSUE TITLE
Unleash service acts multiple times on a `sh.keptn.event.action.triggered`

### ISSUE THREAD

**Issue by johannes-b** [keptn/keptn#7684]
While the bug title names the unleash-service as the subject of investigation, this bug does most likely not relate to the unleash-service but rather the distributor. 

**Environment**

* Keptn Version (`keptn version`): 0.15.0
* Unleash-service: 0.3.2
* Distributor: 0.14.1

```
PS C:\dynatrace\repos\src\github.com\keptn\keptn> kubectl describe deployment unleash-service -n keptn
Name:                   unleash-service
Namespace:              keptn
CreationTimestamp:      Mon, 04 Apr 2022 17:16:34 +0200
Labels:                 <none>
Annotations:            deployment.kubernetes.io/revision: 2
Selector:               run=unleash-service
StrategyType:           RollingUpdate
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
                    app.kubernetes.io/version=0.3.2
                    run=unleash-service
  Containers:
   unleash-service:
    Image:      keptncontrib/unleash-service:0.3.2
    Port:       8080/TCP
    Host Port:  0/TCP
    Environment Variables from:
      unleash  Secret  Optional: true
    Environment:
      EVENTBROKER:            http://localhost:8081/event
      CONFIGURATION_SERVICE:  http://configuration-service:8080
    Mounts:                   <none>
   distributor:
    Image:      keptn/distributor:0.14.1
    Port:       8080/TCP
    Host Port:  0/TCP
    Limits:
      cpu:     250m
      memory:  32Mi
    Requests:
      cpu:     25m
      memory:  16Mi
    Environment:
      PUBSUB_TOPIC:         sh.keptn.>
      PUBSUB_RECIPIENT:     127.0.0.1
      VERSION:               (v1:metadata.labels['app.kubernetes.io/version'])
      K8S_DEPLOYMENT_NAME:   (v1:metadata.labels['app.kubernetes.io/name'])
      K8S_POD_NAME:          (v1:metadata.name)
      K8S_NAMESPACE:         (v1:metadata.namespace)
      K8S_NODE_NAME:         (v1:spec.nodeName)
    Mounts:                 <none>
  Volumes:                  <none>
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Progressing    True    NewReplicaSetAvailable
  Available      True    MinimumReplicasAvailable
OldReplicaSets:  <none>
NewReplicaSet:   unleash-service-6485db6657 (1/1 replicas created)
Events:          <none>
```

**Affected Component**

* [ x ] Distributor / Event Broker

**Describe the bug**

* The `remediation` sequence was triggered. 
* The `unleash-service` reacted on a `sh.keptn.event.action.triggered`, but multiple times: 

![image](https://user-images.githubusercontent.com/729071/167398584-40e3dd76-68a2-480f-baba-8259460c79e7.png)

**Expected behavior**
* Unleash-service acts on the `sh.keptn.event. action.triggered` one time.

**Comment by johannes-b** [comment]
If needed, I can easily reproduce this bug by just sending the sequence trigger again.

**Comment by bacherfl** [comment]
@johannes-b did that happen when the distributor was running inside the same cluster or as a remote execution plane service?

**Comment by bacherfl** [comment]
nevermind, just reproduced the problem in my environment

**Comment by bacherfl** [comment]
I just investigated the problem - here are my findings:

The distributor does seem to work correctly, as it forwards the action.triggered event only once to the unleash service, as can be seen in the following logs:

```
distributor time="2022-05-12T07:02:42Z" level=info msg="Received a message for topic [sh.keptn.event.action.triggered]\n"                                                                                                                                  │
│ distributor time="2022-05-12T07:02:42Z" level=info msg="Sending CloudEvent with ID 47726f5a-4605-46ba-9fad-2eb8165e2701 to 127.0.0.1"  <<<<<<forwarding of event to 127.0.0.1 only happens once                                                                                                                     │
│ distributor time="2022-05-12T07:02:42Z" level=info msg="Received CloudEvent with ID 524d868c-ec75-47d2-8894-443853f64964 - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:42Z" level=info msg="Sent: 524d868c-ec75-47d2-8894-443853f64964, accepted: true"                                                                                                                                        │
│ distributor time="2022-05-12T07:02:42Z" level=info msg="Received CloudEvent with ID 9d9f70ba-b102-4810-84f4-b5fee72c75bb - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:42Z" level=info msg="Received '.finished' event with status 'errored'. Forwarding log message to log ingestion API"                                                                                                     │
│ distributor time="2022-05-12T07:02:42Z" level=info msg="Sent: 9d9f70ba-b102-4810-84f4-b5fee72c75bb, accepted: true"                                                                                                                                        │
│ distributor time="2022-05-12T07:02:43Z" level=info msg="Received CloudEvent with ID da6cff5d-04b8-4d4b-a188-69628b0122b0 - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:43Z" level=info msg="Sent: da6cff5d-04b8-4d4b-a188-69628b0122b0, accepted: true"                                                                                                                                        │
│ distributor time="2022-05-12T07:02:43Z" level=info msg="Received CloudEvent with ID 53f6f165-5655-4dba-aaf5-c465257efcf1 - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:43Z" level=info msg="Sent: 53f6f165-5655-4dba-aaf5-c465257efcf1, accepted: true"                                                                                                                                        │
│ distributor time="2022-05-12T07:02:43Z" level=info msg="Received '.finished' event with status 'errored'. Forwarding log message to log ingestion API"                                                                                                     │
│ distributor time="2022-05-12T07:02:44Z" level=info msg="Received CloudEvent with ID 28823468-a127-4b2d-b004-9a6fe403be27 - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:44Z" level=info msg="Sent: 28823468-a127-4b2d-b004-9a6fe403be27, accepted: true"                                                                                                                                        │
│ distributor time="2022-05-12T07:02:44Z" level=info msg="Received CloudEvent with ID 52d893a6-acf7-4783-b890-a9b6d492b1a6 - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:44Z" level=info msg="Sent: 52d893a6-acf7-4783-b890-a9b6d492b1a6, accepted: true"                                                                                                                                        │
│ distributor time="2022-05-12T07:02:44Z" level=info msg="Received '.finished' event with status 'errored'. Forwarding log message to log ingestion API"                                                                                                     │
│ distributor time="2022-05-12T07:02:46Z" level=info msg="Received CloudEvent with ID 8a9aa307-b2cf-4e7c-a6a3-d14203f2a946 - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:46Z" level=info msg="Sent: 8a9aa307-b2cf-4e7c-a6a3-d14203f2a946, accepted: true"                                                                                                                                        │
│ distributor time="2022-05-12T07:02:46Z" level=info msg="Received CloudEvent with ID c63f86e3-75b7-4b46-8297-996615e9cc3d - Forwarding to Keptn"                                                                                                            │
│ distributor time="2022-05-12T07:02:46Z" level=info msg="Sent: c63f86e3-75b7-4b46-8297-996615e9cc3d, accepted: true"
```

However, as can also be seen in the logs is that the unleash service tries to process the event multiple times. The reason seems to be an undocumented behavior of the CloudEvents library, which leads the unleash service to try to process the same event multiple times if an error is returned by the event handler, as it is the case in the following line:

https://github.com/keptn-contrib/unleash-service/blob/34856576f67965c2a9ff461552c1613a23984c3d/cmd/main.go#L67

In most of the other services we are starting a new goroutine for processing an event and only log errors, which would explain why this problem did not occur in other services as well

**Comment by christian-kreuzberger-dtx** [comment]
I just looked into this, and my findings are the opposite of what Florian provided.

I tried to verify the behaviour of Cloud Events library using this simple Unit Test, and Cloud Events library is only executing once:

```golang
processCloudEventFunc := func(ctx context.Context, event cloudevents.Event) error {
		cnt += 1

		log.Printf("Throwing an error")
		return fmt.Errorf("Some error parsing the CE happened")
	}

	go c.StartReceiver(context.TODO(), processCloudEventFunc)
...
	resp, err := http.Post("http://localhost:8080/", "application/cloudevents+json", responseBody)

	require.Eventuallyf(t, func() bool {
		log.Printf("count=%d\n", cnt)
		return cnt == 1
	}, 60*time.Second, 5*time.Second, "cnt should be 1")
```

I will also try to see if this is coming from `keptn/distributor` or not.

**Comment by christian-kreuzberger-dtx** [comment]
Keptn 0.13.4, keptn/distributor:0.8.4 and unleash-service version 0.3.2: Everything is fine (as in: event is only executed once)
![image](https://user-images.githubusercontent.com/56065213/168020350-16d8b407-307d-43b3-9ef7-355fd0a599ca.png)
Unleash-service and cloud events library are working as expected.

Keptn 0.13.4, keptn/distributor:0.13.4 and unleash-service version 0.3.2: Event is processed (and probably retrieved) multiple times
![image](https://user-images.githubusercontent.com/56065213/168020615-5d3f8757-28bf-417a-800b-b7589e9b5178.png)

unleash-service and cloudevents library are, in the case of an error, returning a HTTP 500. which is also logged by distributor (0.13.4):
```
time="2022-05-12T07:55:05Z" level=error msg="Could not send event for subscription 57d227b9-922d-4066-bd84-d567822c8f60: Failed to send cloudevent: 500: "
```

I was able to verify the same behaviour using distributor 0.11.x and 0.12.x. For me, the rootcause is somehwere hidden in distributor which might be re-sending the Cloud Event in case of an error (which I assume is correct).

**Long story short**, we can fix this in unleash, but this could affect many other services/integrations.

**Comment by bacherfl** [comment]
Just did some further digging, and the retry logic actually comes from there: https://github.com/keptn/go-utils/blob/13539bfe70a95c9947fc1db9de3a25ca3f9af15c/pkg/lib/v0_2_0/events.go#L87 - this sender is also used by the distributor now. However, the event handler in the unleash service should be adapted to just acknowledge the reception of the event, while errors during the execution of the action task should be communicated via .finished events with result/status properties being set appropriately
